not-to-release/sources/answers/20111031103114AA61BW3_ans.xml.conllu

```diff
@@ -273,14 +273,14 @@
 10	some	some	DET	DT	_	12	det	12:det	_
 11	great	great	ADJ	JJ	Degree=Pos	12	amod	12:amod	_
 12	fries	fry	NOUN	NNS	Number=Plur	8	conj	4:obl:for|8:conj:and	_
-13	(	(	PUNCT	-LRB-	_	15	punct	15:punct	SpaceAfter=No
-14	they	they	PRON	PRP	Case=Nom|Number=Plur|Person=3|PronType=Prs	15	nsubj	15:nsubj	_
-15	are	be	VERB	VBP	Mood=Ind|Number=Plur|Person=3|Tense=Pres|VerbForm=Fin	12	parataxis	12:parataxis	_
+13	(	(	PUNCT	-LRB-	_	17	punct	17:punct	SpaceAfter=No
+14	they	they	PRON	PRP	Case=Nom|Number=Plur|Person=3|PronType=Prs	17	nsubj:outer	17:nsubj:outer	_
```

Comment:
So you want to use `:outer` even in cases where there is only one `nsubj` dependent? Won't that be difficult to distinguish in practice? I mean, with two nsubj nodes it's usually not hard to automatically know which is 'outer' in most languages, but if the language doesn't have very clear inf/fin distributions around this (which is possibly most languages in the world, at least the typical 'ural-altaic' type), then we wouldn't be able to get this done via a script.

I thought the main idea was to make sure that multiple nsubj is not an error during validation? If you want to encode phrase scope for subjects, this becomes a much bigger burden on future annotation efforts. I just thought this was a little cosmetic embelishment to distinguish which of two nsubjs is which...

Reply:
The outermost-of-two-overt-subjects interpretation is not what we concluded from the discussion—see the guidelines: https://universaldependencies.org/u/overview/complex-syntax.html#predicate-clauses As I recall the point was that UD cares about clauses, so knowing whether the subject is in the innermost clause or not is valuable even if only one of the two subjects is overt.

Given that the subtype is technically optional to begin with, I suppose it's up to languages how they want to implement it, and the validator will only complain if there are two non-outer subjects present.

Reply:
It's true that UD cares about being a clause head (ccomp, csubj), but not about being a dependent of a clause. 

I can see the appeal of having a way to find subjects that are motivated not by the lexical argument structure of the head, but by a larger construction (which is what I think we are talking about), but I'm wondering if it's worth the price. As long as it's relatively trivial (switch the first of 2 nsubj to `:outer` or the validator will complain), then I think we could get this applied to various TBs and it could have some consistent utility. 

But if we're talking about identifying all such cases even when only one subject is realized, I'm pretty sure that 1. most languages will never implement this, and 2. even in lanugages that do, we will see inconsistent implementations, since it is quite tricky (I don't know for sure whether "they are to travel tomorrow" is outer or not?). So you could see TBs with `:outer` and not know exactly what it means.

So in sum, the question for me is, does this improve UD, or are we just introducing hurdles that will inevitably push the TBs apart in both the label sets and implementation details even when the labels are used? 

I'm not sure if @dan-zeman is reading this so adding him too in case he has some input on the theory or the validator aspects.

Reply:
In terms of other languages being able to implement—if non-zero-copula languages were following the old guidelines they would have had the copula as predicate, right? So I would think it fairly straightforward to convert for those languages.

For English, the not-to-release/tools/outer-subj.sh script this PR adds detects outer subjects *after* the head is reassigned, so it should work for GUM as well (but may produce some false positives with participles tagged as VERB).

"They are to travel tomorrow"—we decided in #310 that "are" is an aux, not a cop, so there is no outer subject.

Reply:
@nschneid is right in that the validator will only complain if it sees multiple non-`:outer` subjects. It will not complain if it sees an `:outer` subject that is not accompanied by a non-`:outer` subject.

From the theoretical point of view, I also agree with @nschneid. The fact that the subject is outer is independent of the presence of the inner subject. Outside English, this can happen even when a finite clause acts as a predicate (the inner subject may be inferrable from the morphology of the verb).

@amir-zeldes is right in that trying to identify all such constructions will not be easy and probably will not be implemented in most treebanks. But it does not sound like a justification of banning it in treebanks where people are willing to do it. Wheteher or not you want to do it in English is something I would like to leave for you guys to decide :-) In general I don't think that a script can solve the multiple-subject errors anyway—maybe it can work for well behaved treebanks that already implemented correctly the previous approach, but most instances of the error I have seen are actually plain errors where no `:outer` subtype should be used.

Reply:
Hm, OK, I see what you're both saying. I'm not sure I completely agree about this relating to a theoretical point of view, because from my perspective, nsubj is just a dependency relation, and in dependencies we don't really know what is inner or outer anywhere because dependencies aren't phrases (I have argued this before about nesting `case` dependents in examples like "jump from under the bed", where we don't mark 'outer'). 

I'm a bit torn about this because I don't want to stand in the way of better, more informative treebanks, and the position of annotating even single 'outer' cases without an internal subject is totally clear and coherent to me. I just get a stomach ache whenever I think about all of the idiosyncracies we introduce into TBs which cause UD to be less homogeneous in the end, and I'm also not a fan of rare relations that inflate confusion matrices when we evaluate systems on something involving deprels.

That said, I can see all the work @nschneid has put into this, and if we have a single script that adds these as post-processing in the same way to EWT and GUM (and potentially the other English TBs?) then maybe it's worth it and not such a big problem.

Can I just make two requests then:

1. I'd like the script to be pure python, since I need the GUM build bot to run cross-platform (no .sh). Since it's udapi this should be fine, right?
2. It looks like the script still requires some manual regexing and one-time steps based on the comments, which will create a frozen corpus that can be manually edited. This doesn't fit for GUM since we will always get new data, and I'd like adding `:outer` to be an automatic last step before edeps. In fact, we may be able to improve the script and prevent false positives down the line, so keeping it 'live' would be better and increases the chance that other, future English TBs will be consistent on this.

From @nschneid 's description it sounds like, since this is running after ccomp fixing anyway, 2. should also be viable, right?

Thanks!

Reply:
In principle I think `:outer` could be added automatically to basic deps without manual fixing (the manual fixes were partly for edeprels, and partly for the cases in #355).

It's a bash script because all the udapy examples I could find were shell commands, but I'm sure @dan-zeman would know how to formulate it as a pure-Python library. Or it could be written as a DepEdit script—just a few rules. (I mainly needed udapy for the other script, be-ccomp.sh, which edits edeps.)

Reply:
> It's a bash script because all the udapy examples I could find were shell commands, but I'm sure @dan-zeman would know how to formulate it as a pure-Python library. Or it could be written as a DepEdit script—just a few rules.

I've always run Udapi from a shell and never from within Python, although I'm quite confident it's possible. @martinpopel will know how :-)

Reply:
OK, I'm not too worried about the details, and it may be simple enough to depedit as well as Nathan said. The main point is that it should be viable to run this as a late, automatic postprocessing step and iteratively improve the implementation. As long as we're not annotating this manually, and English corpora are more or less guaranteed to have the same implementation by applying the same script, then I guess GUM will be on board too. 

I just feel like we're introducing a lot of changes all the time and I really worry about stability, so if we can refrain from adding more deprels for the near future I think that will be a good thing.

Reply:
OK, based on the review, I'd like to see a little higher accuracy before unleashing this on GUM, but as soon as the script works in the form outlined above, I will do a test run and spot check on GUM, just let me know. Thanks!

Reply:
Are there errors you see that are due to the script and not due to the underlying data?

Reply:
That's hard for me to say without knowing the details of how the script works and what issues EWT has in detail. If you're saying the script is basically good to go and can pythonize it, I can give it a go already and see what happens!

Reply:
I believe it is OK. But of course there may be things that pop up due to different conventions in GUM.

Reply:
TBC all the outer-subj.sh script does is to implement these four rules with respect to basic deps:

```
Mark a subject with :outer
      - before another nsubj or csubj (:outer1)
      - before a pre-head expl (:outer2)
      - before a cop dependent of a VERB/AUX (:outer3)
      - before the first of two cop dependents of a predicate (:outer4)
```

The numeric suffixes are just for explainability when you review the data. Of course these should all just be `:outer` in the end.

The reason implementing this for EWT was more complicated than just running this script is that the edeps had to be edited as well.

Reply:
If that's all, then I could easily implement it in a few lines of depedit too... I don't mind doing it, but if so, then I'd like for EWT to use the exact same script, so that there's parity. Would you be OK with switching the last step to depedit for EWT as well?

Reply:
I've already completed and merged the EWT edits. Since EWT is not growing it was a one-time script. For GUM, you could run both scripts just to verify that the output matches.

Reply:
I guess so, but I'm thinking of this as a future investment. It's something I want to bring up tomorrow too maybe, we should really start thinking of consolidating datasets more, so ideally I would like to see this and other things applied to the other English corpora, and then we can think beyond that what we want to do with all of the datasets the validator will flag now that multiple nsubj is strictly forbidden.